feat: Add new `ColorPicker` component

saaaaaally opened the pull request:
Closes #312

![image](https://user-images.githubusercontent.com/9084735/138881523-54fd5266-d69b-4edd-b762-9564d7e9add2.png)
![image](https://user-images.githubusercontent.com/9084735/138881915-a014d5ca-4cb5-4c5f-995d-4384b33bc619.png)

## Checklist

- [x] Add meaningful unit tests for your component (verify that all lines are covered)
- [x] Verify that all existing tests pass
- [x] Add component features demo in Storybook (different stories)
- [x] Approve test images for new stories
- [x] Add screenshots of the key elements of the component

saaaaaally committed:
Create ColorPicker Component

saaaaaally committed:
Add colors to basic color picker

saaaaaally committed:
Merge branch 'main' into saally/color-picker

saaaaaally committed:
Add advanced color picker

saaaaaally committed:
Merge branch 'main' into saally/color-picker

saaaaaally committed:
Set active when clicking a color

saaaaaally committed:
Add onColorClicked handler and basic Color Picker tests

saaaaaally committed:
Update css package

saaaaaally committed:
Made color picker accept list of ReactNode with custom click handler

saaaaaally committed:
Added default Colors list and story with custom click handler

saaaaaally committed:
Added WIthTooltip story

saaaaaally committed:
Removed advanced color picker

saaaaaally committed:
Update color picker tests

saaaaaally committed:
Visual Testing

saaaaaally committed:
Merge branch 'main' into saally/color-picker

saaaaaally committed:
Fixed DefaultColors constant

saaaaaally committed:
Clean up ColorPicker and Color component

saaaaaally committed:
Refactored to accept colors as children only, removed colors prop

saaaaaally committed:
Refactored Color onColorClicked to use extend span props and combined basic and WithCustomClick stories

saaaaaally committed:
Renamed Color to ColorSwatch

saaaaaally committed:
Cleaned up color picker and color swatch and added forward ref

saaaaaally committed:
Clean up color picker and updated tests

saaaaaally committed:
Added arrow key navigation

saaaaaally committed:
Cleaned up ColorSwatch naming and ColorPicker story

saaaaaally committed:
Fixed arrowkey navigation to work dynamically

saaaaaally committed:
Add color swatch to index

saaaaaally committed:
Removed activeColor prop from ColorPicker

saaaaaally committed:
Fix arrow navigation

saaaaaally committed:
Update visual tests

saaaaaally committed:
Merge branch 'main' into saally/color-picker

saaaaaally committed:
Update unit tests

saaaaaally committed:
Refactored arrowkey navigation

saaaaaally committed:
Merge branch 'main' into saally/color-picker

saaaaaally committed:
Clean up ColorPicker and ColorSwatch

saaaaaally committed:
Add ColorSwatch Test

saaaaaally committed:
Updated ColorPicker tests

saaaaaally committed:
Moved DefaultColors list into Story

saaaaaally committed:
Merge branch 'main' into saally/color-picker

saaaaaally committed:
Fixed withTooltip story

saaaaaally committed:
Advanced Color Picker initial push

saaaaaally committed:
Fixed variable naming

saaaaaally committed:
Merge branch 'main' into saally/color-picker

saaaaaally committed:
Merge branch 'saally/color-picker' into saally/color-picker-advanced

saaaaaally committed:
Fixed build error

saaaaaally committed:
Merge branch 'saally/color-picker' into saally/color-picker-advanced

saaaaaally committed:
Clean up

saaaaaally committed:
Merge branch 'saally/color-picker' into saally/color-picker-advanced

saaaaaally committed:
Fix build error

saaaaaally committed:
Add Visual Test

saaaaaally committed:
Improved updateSliderValue and updateSquareValue

saaaaaally committed:
Merge branch 'main' into saally/color-picker

saaaaaally committed:
Merge branch 'saally/color-picker' into saally/color-picker-advanced

saaaaaally committed:
Added HslColor

saaaaaally committed:
Fix build error

saaaaaally committed:
Add update color slider by dragging

saaaaaally committed:
Add update savedColors when adding a color

saaaaaally committed:
Add Color type

saaaaaally committed:
Add calculate new dotColor value using hsv

saaaaaally committed:
Merge branch 'main' into saally/color-picker

saaaaaally committed:
Merge branch 'saally/color-picker' into saally/color-picker-advanced

saaaaaally committed:
Merge remote-tracking branch 'origin/main' into saally/color-picker-advanced

saaaaaally committed:
fix build error

saaaaaally committed:
install tinycolor2

saaaaaally committed:
Added update color on dragging color dot

saaaaaally committed:
Add colorPaletteTitle prop

saaaaaally committed:
Clean up

saaaaaally committed:
Add arrow key navigation for color dot

saaaaaally committed:
Add arrow key navigation for slider

saaaaaally committed:
Improve dragging on slider dot

saaaaaally committed:
Improve dragging on color square

saaaaaally committed:
Merge branch 'main' into saally/color-picker

saaaaaally committed:
Merge branch 'saally/color-picker' into saally/color-picker-advanced

saaaaaally committed:
Set initial square and slider dot values

saaaaaally committed:
Add some advanced color picker tests

saaaaaally committed:
Fixed color swatch not working on its own

saaaaaally committed:
Update visual tests

saaaaaally committed:
Merge branch 'main' into saally/color-picker

saaaaaally committed:
Merge branch 'saally/color-picker' into saally/color-picker-advanced

saaaaaally committed:
Add test for color conversion

saaaaaally committed:
Fixed arrowkey navigation not working with button

bsteinbk committed:
Remove use of tinycolor2 (#374)

saaaaaally committed:
Fixed failing test

veekeys commented on src/core/ColorPicker/ColorPicker.tsx:
```diff
@@ -0,0 +1,546 @@
+  };
+};
+
+export type ColorPickerProps = {
```
Would need to think about the naming, but I feel we could still have colors as children and get rid of `onAdd`. Instead, we could support ability to have that (passing add button together with `ColorSwatch` components as children). That would be more flexible for what ever people might want to have there. Do you guys think it would make sense? @bsteinbk @saaaaaally @mayank99

bsteinbk replied:
@saaaaaally - That sounds OK to me. I plan to get back to making these updates to ColorValue later today or early Monday.

veekeys commented on src/core/utils/color/ColorValue.ts:
```diff
@@ -0,0 +1,709 @@
+  }
+
+  /** Convert this ColorValue to a string in the form "hsva(h,s,v,a)" - i.e hsva(120,50%,50%,0.5). */
+  public toHsvaString(): string {
```
Hmm... Maybe we could save alpha value from the input and if it is defined, return it with all schemes. That would lessen our code and require just to have `toHSVString`, etc...

bsteinbk replied:
What if showAlpha is requested but it is set to 1.... in this case tinycolor2 would just return the hsv()/ rgb() functions and not the hsva/rgba - and that may confuse the user who request to see the alpha value.... and I think that is the same functionality you are requesting.

veekeys commented on src/core/utils/color/ColorValue.ts:
```diff
@@ -0,0 +1,709 @@
+  }
+
+  /** Create a ColorValue from an HsvColor */
+  public static fromHSV(hsv: HsvColor, transparency = 0): ColorValue {
```
Not sure about having separate methods to get `ColorValue` and having constructor with only tbgr support. Could we have just constructor, which could accept any of the supported formats as input and create `ColorValue` object that way? Would be consistent between different color schemes and possibly easier to use.

bsteinbk replied:
used with displaying values. see ColorValue.create for all supported input types.

veekeys commented on src/core/utils/color/ColorValue.ts:
```diff
@@ -0,0 +1,709 @@
+  }
+
+  /** Compute the 0xTTBBGGRR color corresponding to the specified hue, saturation, lightness values. */
+  public static computeTbgrFromHSL(
```
Wouldnt it simplify, if we just ask to create `ColorValue` instance from what ever color input and then have method `toTbgr`? Similar as with other color schemes

veekeys commented on src/core/utils/color/ColorValue.ts:
```diff
@@ -0,0 +1,709 @@
+  }
+
+  /** True if the specified 0xTTBBGGRR color is fully opaque. */
+  public static isOpaque(tbgr: number): boolean {
```
Not sure this is needed.. Pretty simple and users can easily check that on their end. We provide ability to get alpha, so..

veekeys commented on src/core/utils/color/ColorValue.ts:
```diff
@@ -0,0 +1,709 @@
+   * *"#rrbbgg"*
+   * *"blanchedAlmond"* (see possible values from [[ColorByName]]). Case-insensitive.
+   */
+  public static fromString(val: string): ColorValue {
```
Yeah, same here. We could just support all of that in constructor.

bsteinbk replied:
see new ColorValue.create method.  constructor method is now private and compatible with class in imodeljs... its just as easy to use `ColorValue.create()` instead of `new ColorValue ()`

veekeys commented on src/core/utils/color/ColorValue.ts:
```diff
@@ -0,0 +1,709 @@
+export class ColorValue {
+  private readonly _tbgr: number;
+
+  public constructor(tbgr: number) {
```
I think would be great and simplify things a little, if we just accept any format in constructor (rhb, hsv, hsl, stringed of those, tbgr??). Would be also great, if `ColorValue` would use rgba (or hsv?? which is easier @mayank99) as base for all the calculations, as we all are more familiar with those color schemes than tbgr (mb not familiar, but jus used to that order than inversed :D). I think we could have method to get tbgr if needed (even though users could just use rgb for that and inverse).

mayank99 replied:
using `type Color = HslColor | RgbColor | HsvColor | string` from https://github.com/iTwin/iTwinUI-react/pull/346#discussion_r729926621, the constructor would change to:
```ts
public contructor(color: Color) {}
```

not sure which format would be easier for maintaining internal state, but I think rgba would be easier to avoid a full rewrite

saaaaaally committed:
Update css package

saaaaaally committed:
Undo unnecessary changes

saaaaaally committed:
undo slider change

saaaaaally committed:
Update color with new color slider

saaaaaally committed:
Add new ColorBuilderProps and ColorPaletteProps

saaaaaally committed:
Update some tests

bsteinbk commented on src/core/utils/color/ColorValue.ts:
```diff
@@ -0,0 +1,709 @@
+  }
+
+  /** Convert this ColorValue to a 32 bit number representing the 0xTTBBGGRR value */
+  public toJSON(): ColorDefProps {
```
got rid of this and other "confusing" methods.

bsteinbk commented on src/core/utils/color/ColorValue.ts:
```diff
@@ -0,0 +1,709 @@
+  }
+
+  /** Obtain the name of the color in the [[ColorByName]] list associated with the specified 0xTTBBGGRR value, or undefined if no such named color exists. */
+  public static getName(tbgr: number): string | undefined {
```
gone

bsteinbk commented on src/core/utils/color/ColorValue.ts:
```diff
@@ -0,0 +1,709 @@
+  }
+
+  /** Convert this ColorValue to a string in the form "rgba(r,g,b,a)" where color values are decimal digits and a is a fraction */
+  public toRgbaString(): string {
```
remove static and instance methods with same name.

saaaaaally committed:
Add color value input options

saaaaaally committed:
Add update color input when typing

saaaaaally committed:
Merge branch 'main' into saally/color-picker-advanced

saaaaaally committed:
Fix unit test

saaaaaally committed:
Clean up comments and variables

saaaaaally committed:
Merge branch 'main' into saally/color-picker-advanced

saaaaaally committed:
Add button for displaying color types in story

saaaaaally committed:
Move icon & update visual test

bsteinbk committed:
Bsteinbk/colorpicker updates (#387)

saaaaaally committed:
Set negative status for invalid color input

bsteinbk committed:
Fix Hue display in color input panel (#394)

saaaaaally committed:
Clean up and update css

mayank99 committed:
Fix bad merge (#396)

mayank99 committed:
cleanup ColorInputPanel and stories

mayank99 committed:
Make color inputs optional

saaaaaally committed:
Add tests for color picker

saaaaaally committed:
Add color input panel tests

saaaaaally committed:
Fix color dot bugs

saaaaaally committed:
Fix slider bug

saaaaaally committed:
Fix tests

mayank99 committed:
Extract ColorPalette into its own component

saaaaaally committed:
Clean up

saaaaaally committed:
Update tests

saaaaaally committed:
Moved window.css mock to beforeAll

saaaaaally committed:
Update visual tests

saaaaaally committed:
Clean up

mayank99 committed:
random cleanup

mayank99 committed:
Fix tests

mayank99 committed:
Fix imports

saaaaaally committed:
Fix input panel bug

mayank99 committed:
Fix extra onChangeCompleted calls on keydown

saaaaaally committed:
Fix onChangeCompleted not called when mouse is out of document for colorSquare only

mayank99 committed:
Add `setFocus` prop

saaaaaally committed:
Clean up ColorInputPanel

saaaaaally committed:
Removed onColorFormatChanged

saaaaaally committed:
Clean up ColorInputPanel input

mayank99 committed:
Use context to simplify API

`paletteProps` and `builderProps` have been removed. User should now pass `ColorBuilder`, `ColorInputPanel`, and `ColorPalette` as children. This makes the API more intuitive and allows for more flexibility.

Also renamed `onChangeCompleted` to `onChangeComplete`.

mayank99 committed:
Fix visual tests

mayank99 committed:
Fix ColorPalette focus and add more tests

mayank99 committed:
Save input hue in `ColorValue`

bsteinbk reviewed (approved):
Looks good to me.

mayank99 committed:
remove ColorValue comparison in tests

mayank99 committed:
applyHsvColorChange on hue input

mayank99 committed:
Update to latest itwinui-css

mayank99 committed:
Improve story
